### PR TITLE
feat: Add optional attribute tier in Create Virtual Device request for C8000V and C8000V SDWAN

### PR DESCRIPTION
**Pull request by kpdhulipala**
feat: Add optional attribute tier in Create Virtual Device request for C8000V and C8000V SDWAN

### CODE REVIEW

File: equinix/resource_network_device.go [hunk 1]
```diff
@@ -442,6 +444,12 @@ func createNetworkDeviceSchema() map[string]*schema.Schema {
 			ConflictsWith: []string{neDeviceSchemaNames["Secondary"]},
 			Description:   neDeviceDescriptions["DiverseFromDeviceUUID"],
 		},
+		neDeviceSchemaNames["Tier"]: {
+			Type:        schema.TypeInt,
+			ForceNew:    true,
+			Optional:    true,
```
Comment (rling-equinix) [inline equinix/resource_network_device.go:450]: Does the API return tier attribute all the time or just for C8000V device? If the API returns it, we need to add computed attribute here.

File: templates/resources/network_device.md.tmpl [hunk 2]
```diff
@@ -51,6 +54,7 @@ The following arguments are supported:
 * `package_code` - (Required) Device software package code.
 * `version` - (Required) Device software software version.
 * `core_count` - (Required) Number of CPU cores used by device. (**NOTE: Use this field to resize your device. When resizing your HA devices, primary device will be upgraded first. If the upgrade failed, device will be automatically rolled back to the previous state with original core number.**)
+* `tier` - (Optional) Select bandwidth tier for your own license, i.e., `0` or `1` or `2` or `3`. Tiers applicable only for C8000V Autonomous or C8000V SDWAN (controller) device types.
```
Comment (rling-equinix) [inline templates/resources/network_device.md.tmpl:57]: Same here, default value will be applied for C8000V device if tier is not provided. We can put that information here.

File: equinix/resource_network_device.go [hunk 3]
```diff
@@ -1368,6 +1382,7 @@ func updateNetworkDeviceResource(primary *ne.Device, secondary *ne.Device, d *sc
 func flattenNetworkDeviceSecondary(device *ne.Device) interface{} {
 	transformed := make(map[string]interface{})
 	transformed[neDeviceSchemaNames["UUID"]] = device.UUID
+	transformed[neDeviceSchemaNames["Tier"]] = device.Tier
```
Comment (rling-equinix) [inline equinix/resource_network_device.go:1385]: In HA case, does tier apply to both primary and secondary? Can we use different tiers for primary and secondary device?

**Review by rling-equinix — CHANGES_REQUESTED** [review]

File: equinix/resource_network_device.go (hunk 3 above)
Comment (kpdhulipala) [reply equinix/resource_network_device.go:1385]: no, we cant use different for primary and secondary

File: equinix/resource_network_device.go (hunk 1 above)
Comment (kpdhulipala) [reply equinix/resource_network_device.go:450]: it returns only for C8000V

File: examples/resources/equinix_network_device/example_10.tf [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Create C8000V BYOL device with bandwidth tier information
```
Comment (ctreatma) [inline examples/resources/equinix_network_device/example_10.tf:1]: When there are multiple examples for a resource or data source, it's better to use more descriptive file names.  In this case, that might be `examples/resources/equinix_network_device/c800v_byol_with_bandwidth_tier.tf`

File: templates/resources/network_device.md.tmpl [hunk 5]
```diff
@@ -51,6 +55,7 @@ The following arguments are supported:
 * `package_code` - (Required) Device software package code.
 * `version` - (Required) Device software software version.
 * `core_count` - (Required) Number of CPU cores used by device. (**NOTE: Use this field to resize your device. When resizing your HA devices, primary device will be upgraded first. If the upgrade failed, device will be automatically rolled back to the previous state with original core number.**)
+* `tier` - (Optional, conflicts with `throughput`,`throughput_unit` ) Select bandwidth tier for your own license, i.e., `0` or `1` or `2` or `3`. Tiers applicable only for C8000V Autonomous or C8000V SDWAN (controller) device types. If not provided, tier is defaulted to '2'.
```
Comment (ctreatma) [inline templates/resources/network_device.md.tmpl:58]: You could replace this "Argument Reference" section with a call to `{{ .SchemaMarkdown | trimspace }}`, and then the argument description would be generated automatically from the resource code.

File: equinix/data_source_network_device.go [hunk 6]
```diff
@@ -295,6 +295,11 @@ func createDataSourceNetworkDeviceSchema() map[string]*schema.Schema {
 			Computed:    true,
 			Description: neDeviceDescriptions["DiverseFromDeviceUUID"],
 		},
+		neDeviceSchemaNames["Tier"]: {
```
Comment (ctreatma) [inline equinix/data_source_network_device.go:298]: The data source gained a `tier` attribute as well, but I don't see a docs update for that.

**Review by ctreatma — APPROVED** [review]
Added some suggestions around managing docs and examples.  Those aren't binding but I recommend looking into them for the next PR if you don't look into them now.

Review from DRE was required because of the changes to `go.mod` and `go.sum`; those look fine to me so I'm approving from that perspective but any comments/reviews on the code take precedence.

File: examples/resources/equinix_network_device/example_10.tf (hunk 4 above)
Comment (kpdhulipala) [reply examples/resources/equinix_network_device/example_10.tf]: thanks @ctreatma , done

File: equinix/data_source_network_device.go (hunk 6 above)
Comment (kpdhulipala) [reply equinix/data_source_network_device.go:298]: added in data source(terraform-provider-equinix/docs/data-sources/network_device.md)

File: examples/resources/equinix_network_device/example_10.tf (hunk 4 above)
Comment (kpdhulipala) [reply examples/resources/equinix_network_device/example_10.tf]: renamed, thank you

File: templates/resources/network_device.md.tmpl (hunk 5 above)
Comment (kpdhulipala) [reply templates/resources/network_device.md.tmpl:58]: will take  in coming PRs, thank you for inputs

**Comment by ctreatma** [comment]
A unit test is failing:

```
=== RUN   TestNetworkDevice_updateResourceData
    resource_network_device_test.go:141: 
        	Error Trace:	/home/runner/work/terraform-provider-equinix/terraform-provider-equinix/equinix/resource_network_device_test.go:141
        	Error:      	Expected nil, but got: &errors.errorString{s:"error reading Secondary: Invalid address to set: []string{\"secondary_device\", \"0\", \"tier\"}"}
        	Test:       	TestNetworkDevice_updateResourceData
        	Messages:   	Update of resource data does not return error
--- FAIL: TestNetworkDevice_updateResourceData (0.00s)
```

File: equinix/resource_network_device.go [hunk 7]
```diff
@@ -119,6 +120,7 @@ var neDeviceDescriptions = map[string]string{
 	"ProjectID":             "The unique identifier of Project Resource to which device is scoped to",
 	"DiverseFromDeviceUUID": "Unique ID of an existing device",
 	"DiverseFromDeviceName": "Diverse Device Name of an existing device",
+	"Tier":                  "bandwidth tiers",
```
[the remark this answers is not in the document]
Comment (kpdhulipala) [reply equinix/resource_network_device.go]: done

File: docs/resources/network_device.md [hunk 8]
```diff
@@ -406,8 +436,9 @@ The following arguments are supported:
 * `metro_code` - (Required) Device location metro code.
 * `hostname` - (Optional) Device hostname prefix.
 * `package_code` - (Required) Device software package code.
-* `version` - (Required) Device software software version.
+* `version` - (Required) Device software version.
 * `core_count` - (Required) Number of CPU cores used by device. (**NOTE: Use this field to resize your device. When resizing your HA devices, primary device will be upgraded first. If the upgrade failed, device will be automatically rolled back to the previous state with original core number.**)
+* `tier` - (Optional) Select bandwidth tier for your own license, i.e., `0` or `1` or `2` or `3`. Tiers applicable only for C8000V Autonomous or C8000V SDWAN (controller) device types.
```
[the remark this answers is not in the document]
Comment (kpdhulipala) [reply docs/resources/network_device.md]: done

File: templates/resources/network_device.md.tmpl (hunk 2 above)
Comment (kpdhulipala) [reply templates/resources/network_device.md.tmpl]: done

**Review by rling-equinix — APPROVED** [review]
Lgtm.

**Review by shyamchalla — APPROVED** [review]

**Review by ctreatma — APPROVED** [review]

**Pull request merged**